### ISSUE TITLE
Results

### ISSUE THREAD

**Issue by ZhouCX117** [yolu1055/conditional-glow#3]
Hi! Thanks for your nice work. I train the model following your readme. But I get all-black images. Could you help me ?

**Comment by yolu1055** [comment]
Hi,

Thank you for your interests. Could you provide more details about your experiment results? For example, are the samples of training set and valid set all black? Are the inverse images black? Did you see the NLL decrease during training? Did you see the samples becoming better during training? Thank you.

Best,

**Comment by yolu1055** [comment]
> Hi! Thanks for your nice work. I train the model following your readme. But I get all-black images. Could you help me ?

**Comment by ZhouCX117** [comment]
Hi, the samples of the training set and validation set are all black ( which are saved in the directory "images" and "valid samples"). The NLL loss during training and validation is all decresing. The samples are all black from the beginning to the end.
![image](https://user-images.githubusercontent.com/79554332/164166713-9d5de624-e625-493c-9824-d29372840c34.png)
The NLL loss decreases from 31 to 1 and keeps around 1.
![image](https://user-images.githubusercontent.com/79554332/164166876-d0545306-f082-4ab0-86dc-e7ec3d9967a8.png)
The validation NLL loss decreases from 22 to 1 and keeps around 1.

**Comment by ZhouCX117** [comment]
Hi, I have read all codes. Why the results are all black. The first row of the pictures is the ground truth. And I can't fully understand the codes `label_i = skimage.color.label2rgb(seg[i], colors=colors)#根据标签值对图片进行着色
        label_i = skimage.util.img_as_ubyte(label_i)` .
If possible, can you check the code and do me a favor? Thanks a lot! I appreciate your answer~

**Comment by ZhouCX117** [comment]
Also, I can't fully understand the "Conv2dNormy" and "Conv2dZerosy" in the affine layer's f module. Why do we need an actnorm layer in the Conv2dNormy layer? And why do we need (output+b)*exp(s*3) in the ConvZerosy layer?

Looking forward to your reply！ Have a good day!

**Comment by yolu1055** [comment]
Hi,

Sorry for the late response. I am too busy these days. For your questions, 

1. the label2rgb is used for converting  gray scale images to rgb images for visualization, and the img_as_ubyte is its reversed operation. This is because the label y is a binary matrix, i.e., each cell y_{i,j} = {0,1}. It can be seen as a gray scale image with each pixel's value is 0 or 1. To visualize the labels as images, we use the label2rgb function.

2. Those functions in Conv2dNormy and Conv2dZerosy are tricks for stabilizing training and improving model performance. Those functions are adapted from the original implementation of Glow.

3. I will re-run the code to check the errors when I am available. But I am sorry that I may not have time to do it in recent days.

Best,

**Comment by ZhouCX117** [comment]
@yolu1055 Thanks for your reply, waiting for your good news！

**Comment by yolu1055** [comment]
Hi, Sorry for the late reply. I re-run the code on the horse dataset, and the results are correct. I think after 1,000 iterations, you should be able to see the samples in the ```/valid_samples``` folder are becoming better and better. After 10,000 iterations, you should be able to see pretty good samples. You don't need to tune the hyper-parameters, and just need to set the input and output directories. Let me know if there are any issues.

**Comment by alephpi** [comment]
OK, the black issue was solved in #7, but the nll loss doesn't drop to near 0 as reported in the paper, can you verify? Particularly I notice that your code has a commit of bug fix eb13fa7f77a46eb1d0adcdfe0f9ea46629d7a153, where the loss is mistakenly computed, can you confirm if the results are evaluated after you fix that bug? I'm looking forward to your reply.

**Comment by yolu1055** [comment]
Hi @alephpi , I think the NLL difference is because in the plots, I use Gaussian noise to dequantize labels, and in the Github code, I use Uniform noise. That is, at this [line](https://github.com/yolu1055/conditional-glow/blob/703a3fea55d1ecb09eed0d19c78ffd9121a3ebad/datasets.py#L17), you can try `x = x + torch.normal(mean=torch.zeros_like(x), std=torch.ones_like(x) * 0.3)`. The Gaussian noise will make the peak of `p(y|x)` sharper, so the NLL is lower. 

I tried both methods in my experiments, but I only reported one in the paper. Sorry for the confusion caused.

**Comment by alephpi** [comment]
> Hi @alephpi , I think the NLL difference is because in the plots, I use Gaussian noise to dequantize labels, and in the Github code, I use Uniform noise. That is, at this [line](https://github.com/yolu1055/conditional-glow/blob/703a3fea55d1ecb09eed0d19c78ffd9121a3ebad/datasets.py#L17), you can try `x = x + torch.normal(mean=torch.zeros_like(x), std=torch.ones_like(x) * 0.3)`. The Gaussian noise will make the peak of `p(y|x)` sharper, so the NLL is lower.
> 
> I tried both methods in my experiments, but I only reported one in the paper. Sorry for the confusion caused.

Thanks. But with std=0.3 the nll loss is still around 1.3, I change it to 0.15(to make sure the span of $3\sigma$ is around $[0.5,0.5]$) and it reduces to ~0.3. But I have one more question, does it means the smaller std we choose the better the result is, is there any price for doing it?

**Comment by yolu1055** [comment]
This is a great question. Dequantization is essentially a method to alleviate model overfitting. When we choose noise with large variance, e.g., uniform distribution, Gaussian distribution with large std, we will get worse NLL, but the models are less likely to get overfitting. When we choose noise with small variance, the probability mass will easily get accumulated to the peak around the true label, so we will have better NLL. But since the data is less diverse, the model is easier to get overfitting. So there is a tradeoff between better NLL and model overfitting, which depends on the problem. 

In my opinion, compared to other metrics, e.g., accuracy, IoU, F1, NLL is not a reliable metric for supervised learning. As you have seen that NLL is easily impacted by some other factors, e.g., the dequantization method. So we only use NLL as a reference metric in the paper.
